modules/rostests/kmtests/rtl/RtlStrSafe.c

```diff
@@ -0,0 +1,151 @@
+ * PROGRAMMER:      Hern�n Di Pietro <hernan.di.pietro@gmail.com>
+ */
+
+#define KMT_EMULATE_KERNEL
+#include <kmt_test.h>
+#include <ntstrsafe.h>
+#include <ntdef.h>
+#include <ndk/rtlfuncs.h>
+
+#define TESTAPI static void
+
+TESTAPI
+Test_RtlUnicodeStringPrintf()
+{
+    WCHAR Buffer[1024];
+    WCHAR BufferSmall[2];
+    WCHAR BufferSmall2[7];
+    UNICODE_STRING UsString;
+    WCHAR FormatStringInts[] = L"%d %d %d";
+    WCHAR FormatStringStrs[] = L"%s %s %s";
+
+    /* No zeros (Don't assume UNICODE_STRINGS are NULL terminated) */
+
+    RtlFillMemory(Buffer, sizeof(Buffer), 0xAA);
+    RtlFillMemory(BufferSmall, sizeof(BufferSmall), 0xAA);
+    RtlFillMemory(BufferSmall2, sizeof(BufferSmall2), 0xAA);
+
+    /* STATUS_SUCCESS test */
+
+    UsString.Buffer = Buffer;
+    UsString.Length = 0;
+    UsString.MaximumLength = sizeof(Buffer);
+
+    const WCHAR Result[] = L"1 2 3";
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, FormatStringInts, 1, 2, 3), STATUS_SUCCESS);
+    ok_eq_uint(UsString.Length, sizeof(Result) - sizeof(WCHAR));
+    ok_eq_uint(UsString.MaximumLength, sizeof(Buffer));
+    ok_eq_wchar(UsString.Buffer[0], L'1');
+    ok_eq_wchar(UsString.Buffer[1], L' ');
+    ok_eq_wchar(UsString.Buffer[2], L'2');
+    ok_eq_wchar(UsString.Buffer[3], L' ');
+    ok_eq_wchar(UsString.Buffer[4], L'3');
+    ok_eq_wchar(UsString.Buffer[5], (WCHAR) 0);
+    
+    /* STATUS_BUFFER_OVERFLOW tests */
+
+    UsString.Buffer = BufferSmall;
+    UsString.Length = 0;
+    UsString.MaximumLength = sizeof(BufferSmall);
+    
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, FormatStringStrs, L"AAA", L"BBB", L"CCC"), STATUS_BUFFER_OVERFLOW);
+    ok_eq_uint(UsString.Length, UsString.MaximumLength);
+    ok_eq_char(UsString.Buffer[0], L'A');
+    ok_eq_char(UsString.Buffer[1], (WCHAR)0);
+
+    UsString.Buffer = BufferSmall2;
+    UsString.Length = 0;
+    UsString.MaximumLength = sizeof(BufferSmall2);
+
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, FormatStringStrs, L"0123", L"4567", L"89AB"), STATUS_BUFFER_OVERFLOW);
+    ok_eq_uint(UsString.Length, UsString.MaximumLength);
+    ok_eq_char(UsString.Buffer[0], L'0');
+    ok_eq_char(UsString.Buffer[1], L'1');
+    ok_eq_char(UsString.Buffer[2], L'2');
+    ok_eq_char(UsString.Buffer[3], L'3');
+    ok_eq_char(UsString.Buffer[4], L' ');
+    ok_eq_char(UsString.Buffer[5], L'4');
+    ok_eq_char(UsString.Buffer[6], (WCHAR) 0);
+
+    ///* STATUS_INVALID_PARAMETER tests */
+
+    ok_eq_hex(RtlUnicodeStringPrintf(NULL, FormatStringStrs, L"AAA", L"BBB", L"CCC"), STATUS_INVALID_PARAMETER);
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, NULL, L"AAA", L"BBB", L"CCC"), STATUS_INVALID_PARAMETER);
+
+    UNICODE_STRING UsStringNull;
+    UsStringNull.Buffer = (PWCH)NULL;
+    UsStringNull.Length = 0;
+    UsStringNull.MaximumLength = 0;
+    ok_eq_bool(RtlUnicodeStringPrintf(&UsStringNull, FormatStringStrs, L"AAA", L"BBB", L"CCC"), STATUS_INVALID_PARAMETER);
```

Comment:
Considering these tests run in kernel mode, I'd wrap them in SEH.
For example: https://git.reactos.org/?p=reactos.git;a=blob;f=dll/directx/ddraw/Ddraw/GetDeviceIdentifier.c;hb=fcf83315dccd67cd9636f9ae08868104a420a3dd#l33
Or, specifically for kmtests: https://git.reactos.org/?p=reactos.git;a=blob;f=modules/rostests/kmtests/fltmgr/fltmgr_register/fltmgr_register.c;hb=fcf83315dccd67cd9636f9ae08868104a420a3dd#l56